java/src/jmri/jmrit/operations/routes/tools/ImportRoutes.java

```diff
@@ -0,0 +1,153 @@
+            }
+
+            if (!headerFound && inputLine[FIELD_ROUTE_NAME].equals(Bundle.getMessage("Route"))) {
+                headerFound = true;
+                int elementNum = 0;
+                for (String lineElement : inputLine) {
+                    log.debug("Header {} is: {}", elementNum++, lineElement);
+                }
+                continue; // skip header
+            }
+            if (inputLine.length < 1) {
+                log.debug("Skipping blank line");
+                continue;
+            }
+            if (inputLine.length == 3) {
+                log.debug("Found start of route: {}", inputLine[FIELD_ROUTE_NAME]);
+                route = routeManager.getRouteByName(inputLine[FIELD_ROUTE_NAME]);
+                if (route != null) {
+                    log.info("Route: {} already exists, will not import!", inputLine[FIELD_ROUTE_NAME]);
+                    route = null;
+                } else {
+                    routesAdded++;
+                    route = routeManager.newRoute(inputLine[FIELD_ROUTE_NAME]);
+                    log.info("Creating route: {}", inputLine[FIELD_ROUTE_NAME]);
+                    route.setComment(inputLine[FIELD_ROUTE_COMMENT]);
+                }
+            }
+            if (route != null && inputLine.length == 15) {
+                log.debug("Adding route location: {}", inputLine[FIELD_ROUTE_LOCATION_NAME]);
+                Location location = locationManager.getLocationByName(inputLine[FIELD_ROUTE_LOCATION_NAME]);
+                RouteLocation rl = route.addLocation(location);
+                rl.setTrainDirection(Setup.getDirectionInt(inputLine[FIELD_ROUTE_LOCATION_DIRECTION]));
+                rl.setMaxCarMoves(Integer.parseInt(inputLine[FIELD_ROUTE_LOCATION_MOVES]));
+                rl.setRandomControl(inputLine[FIELD_ROUTE_LOCATION_RANDOM]);
+                rl.setPickUpAllowed(inputLine[FIELD_ROUTE_LOCATION_PICKUP_ALLOWED].equals(Bundle.getMessage("yes")));
+                rl.setDropAllowed(inputLine[FIELD_ROUTE_LOCATION_DROP_ALLOWED].equals(Bundle.getMessage("yes")));
+                rl.setWait(Integer.parseInt(inputLine[FIELD_ROUTE_LOCATION_TRAVEL]) - Setup.getTravelTime());
+                rl.setDepartureTime(inputLine[FIELD_ROUTE_LOCATION_DEPARTURE_TIME]);
+                rl.setMaxTrainLength(Integer.parseInt(inputLine[FIELD_ROUTE_LOCATION_TRAIN_LENGTH]));
+                rl.setGrade(Double.parseDouble(inputLine[FIELD_ROUTE_LOCATION_GRADE]));
+                rl.setTrainIconX(Integer.parseInt(inputLine[FIELD_ROUTE_LOCATION_ICON_X]));
+                rl.setTrainIconY(Integer.parseInt(inputLine[FIELD_ROUTE_LOCATION_ICON_Y]));
+                rl.setComment(inputLine[FIELD_ROUTE_LOCATION_COMMENT]);
+                rl.setCommentTextColor(inputLine[FIELD_ROUTE_LOCATION_COMMENT_COLOR]);
+            }
+        }
+
+        if (importOkay) {
+            JmriJOptionPane.showMessageDialog(null, Bundle.getMessage("ImportRoutesAdded", routesAdded),
```

Comment:
JOptionPane, which I suppose JmriJOptionPane is based on, is not thread safe. So you need to ensure that the call to JmriJOptionPane is on the Swing thread (GUI thread).

https://docs.oracle.com/javase/8/docs/api/javax/swing/JOptionPane.html

Quote from the link above:
Warning: Swing is not thread safe. For more information see Swing's Threading Policy.

Reply:
The import is running on its own thread:

   @Override
    public void actionPerformed(ActionEvent e) {
        Thread mb = new ImportRoutes();
        mb.setName("Import Routes"); // NOI18N
        mb.start();
    }

I could instead run the import on the GUI thread.

Reply:
No need to run the import on the GUI thread. But the call to JmriJOptionPane must be on the GUI thread.

```
ThreadingUtil.runOnGUI( ()->{
    JmriJOptionPane.showMessageDialog(null, Bundle.getMessage("ImportRoutesAdded", routesAdded),
    ...
});
```

`runOnGUI()` will wait for the call to `JmriJOptionPane.showMessageDialog()` to finish before it returns.

Reply:
You said "But the call to JmriJOptionPane must be on the GUI thread."

I think I understand thread safe operation, what do you believe running on the GUI does for the showMessageDialog pane?  The importRoutes thread waits for the dialog to finish.

Reply:
> You said "But the call to JmriJOptionPane must be on the GUI thread."
> 
> I think I understand thread safe operation, what do you believe running on the GUI does for the showMessageDialog pane? The importRoutes thread waits for the dialog to finish.

I believe that the `JDialog.setVisible(true)` call must be run on the Swing (GUI) thread. Otherwise strange things might happen. And the showMessageDialog() method calls JDialog.setVisible(true).